File: LeetCode/2-add_two_numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        head, tail, carry = None, None, 0
        while l1 and l2:
            tmp = l1.val + l2.val + carry
            node = ListNode(tmp % 10)
            carry = tmp // 10
            if not tail:
                head = tail = node
            else:
                tail.next = node
                tail = node
            l1, l2 = l1.next, l2.next
        while l1:
            tmp = l1.val + carry
            tail.next = ListNode(tmp % 10)
            tail = tail.next
            carry = tmp // 10
            l1 = l1.next
        while l2:
            tmp = l2.val + carry
            tail.next = ListNode(tmp % 10)
            tail = tail.next
            carry = tmp // 10
            l2 = l2.next
        while carry > 0:
            tail.next = ListNode(carry % 10)
            tail = tail.next
            carry = carry // 10
        return head

File: LeetCode/2-add_two_numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

struct ListNode;

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2)
    {
        ListNode *head = nullptr, *tail = nullptr;
        int carry = 0;

        while (l1 && l2) {
            int tmp = carry + l1->val + l2->val;
            ListNode* node = new ListNode(tmp % 10);
            if (head == nullptr) {
                head = tail = node;
            } else {
                tail->next = node;
                tail = node;
            }
            carry = tmp / 10;
            l1 = l1->next;
            l2 = l2->next;
        }

        if (l2)
            l1 = l2;

        while (l1) {
            int tmp = carry + l1->val;
            ListNode* node = new ListNode(tmp % 10);
            if (head == nullptr) {
                head = tail = node;
            } else {
                tail->next = node;
                tail = node;
            }
            carry = tmp / 10;
            l1 = l1->next;
        }

        while (carry > 0) {
            int tmp = carry;
            ListNode* node = new ListNode(tmp % 10);
            if (head == nullptr) {
                head = tail = node;
            } else {
                tail->next = node;
                tail = node;
            }
            carry = tmp / 10;
        }

        return head;
    }
};
